toolsrc/src/tls12-download.c

```diff
@@ -0,0 +1,263 @@
+#include <Windows.h>
+#include <process.h>
+#include <winhttp.h>
+/*
+ * This program must be as small as possible, because it is committed in binary form to the
+ * vcpkg github repo to enable downloading the main vcpkg program on Windows 7, where TLS 1.2 is
+ * unavailable to PowerShell.
+ * To that end it avoids using C runtime functions (beyond the vcruntime ones the compiler
+ * injects itself).
+ * (In testing as of 2021-01-07, this version that doesn't link with the CRT is ~8kb, whereas a
+ * hello world program that does link with the CRT is ~300kb)
+ */
+
+void __declspec(noreturn) win32_abort() { TerminateProcess(GetCurrentProcess(), 3); }
+
+size_t wide_length(const wchar_t* str)
+{
+    size_t answer = 0;
+    while (*str)
+    {
+        ++answer;
+        ++str;
+    }
+    return answer;
+}
+
+void write_message(const HANDLE hStdOut, const wchar_t* msg)
+{
+    size_t wcharsToWrite = wide_length(msg);
+    if (wcharsToWrite == 0)
+    {
+        return;
+    }
+
+    if (wcharsToWrite > 65535)
+    {
+        win32_abort();
+    }
+
+    if (WriteConsoleW(hStdOut, msg, wcharsToWrite, 0, 0))
+    {
+        return;
+    }
+
+    // this happens if output has been redirected
+    int narrowChars = WideCharToMultiByte(CP_ACP, 0, msg, (int)wcharsToWrite, 0, 0, 0, 0);
+    if (narrowChars == 0)
+    {
+        win32_abort();
+    }
+
+    char* narrowBuffer = HeapAlloc(GetProcessHeap(), 0, (size_t)narrowChars);
+    if (WideCharToMultiByte(CP_ACP, 0, msg, (int)wcharsToWrite, narrowBuffer, narrowChars, 0, 0) == 0)
+    {
+        win32_abort();
+    }
+
+    while (narrowChars != 0)
+    {
+        DWORD charsWritten;
+        if (!WriteFile(hStdOut, narrowBuffer, (DWORD)narrowChars, &charsWritten, 0))
+        {
+            win32_abort();
+        }
+
+        narrowChars -= (int)charsWritten;
+    }
+
+    if (!HeapFree(GetProcessHeap(), 0, narrowBuffer))
+    {
+        win32_abort();
+    }
+}
+
+void write_number(const HANDLE hStdOut, DWORD number)
+{
+    wchar_t buffer[11]; // 4294967295\0
+    wchar_t* cursor = buffer + 11;
+    *--cursor = L'\0';
+    if (number == 0)
+    {
+        *--cursor = L'0';
+    }
+    else
+    {
+        do
+        {
+            *--cursor = L'0' + number % 10;
+            number /= 10;
+        } while (number != 0);
+    }
+
+    write_message(hStdOut, cursor);
+}
+
+void report_api_failure(const HANDLE hStdOut, const wchar_t* api_name)

```

Comment:
```suggestion
void __declspec(noreturn) report_api_failure(const HANDLE hStdOut, const wchar_t* api_name)
```
It also wasn't _immediately_ obvious to me at the call site that this aborted; it would be nice to have "abort" or "exit" or "terminate" in the function name.

Reply:
Renamed to `abort_api_failure`.